libsolidity/lsp/LanguageServer.cpp

```diff
@@ -0,0 +1,746 @@
+
+	auto const sourceNode = findASTNode(_documentPosition.position, sourceName);
+	if (!sourceNode)
+	{
+		fprintf(stderr, "AST node not found\n");
+		// error(_params.requestId, ErrorCode::InvalidParams, "Symbol not found.");
+		return {};
+	}
+
+	auto output = vector<::lsp::Location>{};
+	if (auto const sourceIdentifier = dynamic_cast<Identifier const*>(sourceNode); sourceIdentifier != nullptr)
+	{
+		auto const sourceName = _documentPosition.uri.substr(7); // strip "file://"
+		frontend::SourceUnit const& sourceUnit = m_compilerStack->ast(sourceName);
+
+		if (auto decl = sourceIdentifier->annotation().referencedDeclaration; decl)
+			findAllReferences(decl, sourceUnit, _documentPosition.uri, output);
+
+		for (auto const decl: sourceIdentifier->annotation().candidateDeclarations)
+			findAllReferences(decl, sourceUnit, _documentPosition.uri, output);
+	}
+	else if (auto const varDecl = dynamic_cast<VariableDeclaration const*>(sourceNode); varDecl != nullptr)
+	{
+		fprintf(stderr, "AST node is vardecl\n");
+		auto const sourceName = _documentPosition.uri.substr(7); // strip "file://"
+		frontend::SourceUnit const& sourceUnit = m_compilerStack->ast(sourceName);
+		findAllReferences(varDecl, sourceUnit, _documentPosition.uri, output);
+	}
+	else if (auto const memberAccess = dynamic_cast<MemberAccess const*>(sourceNode); memberAccess != nullptr)
+	{
+		auto const sourceName = _documentPosition.uri.substr(7); // strip "file://"
+		frontend::SourceUnit const& sourceUnit = m_compilerStack->ast(sourceName);
+		findAllReferences(varDecl, sourceUnit, _documentPosition.uri, output);
+	}
+	else
+	{
+		fprintf(stderr, "not an identifier\n");
+	}
+	return output;
+}
+
+vector<::lsp::DocumentHighlight> LanguageServer::semanticHighlight(::lsp::DocumentPosition _documentPosition)
+{
+	fprintf(stderr, "DocumentHighlightParams: %s:%d:%d\n",
+		_documentPosition.uri.c_str(),
+		_documentPosition.position.line,
+		_documentPosition.position.column
+	);
+
+	auto const file = m_vfs.find(_documentPosition.uri);
+	if (!file)
+	{
+		// reply(_params.requestId, output);
+		// error(_documentPosition.requestId, ErrorCode::RequestCancelled, "not implemented yet.");
+		return {};
+	}
+
+	compile(*file);
+	solAssert(m_compilerStack.get() != nullptr, "");
+
+	auto const sourceName = file->uri().substr(7); // strip "file://"
+
+	auto const sourceNode = findASTNode(_documentPosition.position, sourceName);
+	if (!sourceNode)
+	{
+		fprintf(stderr, "semanticHighlight: AST node not found\n");
+		// error(_documentPosition.requestId, ErrorCode::InvalidParams, "Symbol not found.");
+		return {};
+	}
+
+	fprintf(stderr, "semanticHighlight: Source Node(%s): %s\n", typeid(*sourceNode).name(), sourceNode->location().text().c_str());
+
+	auto output = vector<::lsp::DocumentHighlight>{};
+
+	// XXX we could make this if/else chain a SimpleASTVisitor
+	if (auto const importDirective = dynamic_cast<ImportDirective const*>(sourceNode); importDirective != nullptr)
```

Comment:
```suggestion
	if (auto const importDirective = dynamic_cast<ImportDirective const*>(sourceNode))
```

Reply:
@chriseth why do you propose the the `*` here? I know that I have to explicitly qualify reference if i need the reference, but with pointers I don't think so, or do you propose for expressiveness?

Reply:
This is our coding style.